File: test/integration/python/aio/publisher.py
import asyncio
import sys
import argparse

from aiostomp import AioStomp

sys.path.append('gen_py_asyncio')
sys.path.append('..')

from frugal.context import FContext
from frugal.provider import FScopeProvider

from frugal.aio.transport import (
    FStompPublisherTransportFactory,
    FStompSubscriberTransportFactory
)

from common.utils import *


async def main():
    global response_received
    parser = argparse.ArgumentParser(
        description='Run a python asyncio stomp publisher')
    parser.add_argument('--port', dest='port', default='9090')
    parser.add_argument('--protocol', dest='protocol_type', default="binary",
                        choices="binary, compact, json")
    parser.add_argument('--transport', dest='transport_type',
                        default=ACTIVEMQ_NAME, choices="activemq")
    args = parser.parse_args()

    protocol_factory = get_protocol_factory(args.protocol_type)

    if args.transport_type == ACTIVEMQ_NAME:
        stomp_client = AioStomp('localhost', 61613)
        await stomp_client.connect()

        pub_transport_factory = FStompPublisherTransportFactory(stomp_client)
        sub_transport_factory = FStompSubscriberTransportFactory(stomp_client)
        provider = FScopeProvider(
            pub_transport_factory, sub_transport_factory, protocol_factory)
        publisher = EventsPublisher(provider)
    else:
        print(
            "Unknown transport type: {type}".format(type=args.transport_type))
        sys.exit(1)

    await publisher.open()

    def subscribe_handler(context, event):
        print("Response received {}".format(event))
        global response_received
        if context:
            response_received = True

    # Subscribe to response
    preamble = "foo"
    ramble = "bar"
    subscriber = EventsSubscriber(provider)
    await subscriber.subscribe_EventCreated(preamble, ramble, "response",
                                            "{}".format(port),
                                            subscribe_handler)

    event = Event(Message="Sending Call")
    context = FContext("Call")
    context.set_request_header(PREAMBLE_HEADER, preamble)
    context.set_request_header(RAMBLE_HEADER, ramble)
    print("Publishing...")
    await publisher.publish_EventCreated(context, preamble, ramble, "call",
                                         "{}".format(port), event)

    # Loop with sleep interval. Fail if not received within 3 seconds
    total_time = 0
    interval = 0.1
    while total_time < 3:
        if response_received:
            break
        else:
            await asyncio.sleep(interval)
            total_time += interval

    if not response_received:
        print("Pub/Sub response timed out!")
        exit(1)

    await publisher.close()
    exit(0)


if __name__ == '__main__':
    io_loop = asyncio.get_event_loop()
    io_loop.run_until_complete(main())

File: test/integration/go/common/publisher.go
/*
 * Copyright 2017 Workiva
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package common

import (
	"fmt"
	"log"
	"time"

	"git.apache.org/thrift.git/lib/go/thrift"

	"github.com/Workiva/frugal/lib/go"
	"github.com/Workiva/frugal/test/integration/go/gen/frugaltest"
)

func StartPublisher(
	host string,
	port int64,
	transport string,
	protocol string,
	pubSub chan bool,
	sent chan bool) error {

	var protocolFactory thrift.TProtocolFactory
	switch protocol {
	case "compact":
		protocolFactory = thrift.NewTCompactProtocolFactory()
	case "json":
		protocolFactory = thrift.NewTJSONProtocolFactory()
	case "binary":
		protocolFactory = thrift.NewTBinaryProtocolFactoryDefault()
	default:
		return fmt.Errorf("Invalid protocol specified %s", protocol)
	}

	go func() {
		<-pubSub
		var pfactory frugal.FPublisherTransportFactory
		var sfactory frugal.FSubscriberTransportFactory

		switch transport {
		case ActiveMqName:
			stompConn := getStompConn()
			pfactory = frugal.NewFStompPublisherTransportFactory(stompConn, 32*1024*1024, "")
			sfactory = frugal.NewFStompSubscriberTransportFactory(stompConn, "", false)
		default:
			panic(fmt.Errorf("invalid transport specified %s", transport))
		}

		provider := frugal.NewFScopeProvider(pfactory, sfactory, frugal.NewFProtocolFactory(protocolFactory))
		publisher := frugaltest.NewEventsPublisher(provider)

		if err := publisher.Open(); err != nil {
			panic(err)
		}
		defer publisher.Close()

		// Start Subscription, pass timeout
		resp := make(chan bool)
		subscriber := frugaltest.NewEventsSubscriber(provider)
		preamble := "foo"
		ramble := "bar"
		// TODO: Document SubscribeEventCreated "user" cannot contain spaces
		_, err := subscriber.SubscribeEventCreated(preamble, ramble, "response", fmt.Sprintf("%d", port), func(ctx frugal.FContext, e *frugaltest.Event) {
			fmt.Printf(" Response received %v\n", e)
			close(resp)
		})
		if err != nil {
			panic(err)
		}
		ctx := frugal.NewFContext("Call")
		ctx.AddRequestHeader(preambleHeader, preamble)
		ctx.AddRequestHeader(rambleHeader, ramble)
		event := &frugaltest.Event{Message: "Sending call"}
		fmt.Print("Publishing... ")
		if err := publisher.PublishEventCreated(ctx, preamble, ramble, "call", fmt.Sprintf("%d", port), event); err != nil {
			panic(err)
		}

		timeout := time.After(time.Second * 3)

		select {
		case <-resp: // Response received is logged in the subscribe
		case <-timeout:
			log.Fatal("Pub/Sub response timed out!")
		}
		close(sent)
	}()

	return nil
}
